src/vs/workbench/contrib/scm/common/scmService.ts

```diff
@@ -75,17 +71,107 @@ class SCMInput implements ISCMInput {
 		this._validateInput = validateInput;
 		this._onDidChangeValidateInput.fire();
 	}
-
+	private index: any;
 	private readonly _onDidChangeValidateInput = new Emitter<void>();
 	readonly onDidChangeValidateInput: Event<void> = this._onDidChangeValidateInput.event;
-
 	constructor(
 		readonly repository: ISCMRepository,
-		@IStorageService private storageService: IStorageService
+		@IStorageService private storageService: IStorageService,
+		readonly history: string[]
 	) {
-		if (this.repository.provider.rootUri) {
-			const key = `scm/input:${this.repository.provider.label}:${this.repository.provider.rootUri.path}`;
-			this._value = this.storageService.get(key, StorageScope.WORKSPACE, '');
+		this.index = 0;
+		this.history = [];
+		let root = this.repository.provider.rootUri;
+		if (root) {
+			const key = `scm/input:${this.repository.provider.label}:${root.path}:latest`;
+			let result = this.storageService.get(key, StorageScope.WORKSPACE);
+			if (result) {
+				this._value = result;
+			} else {
+				this._value = "";
+			}
+			const historyKey = `scm/input:${this.repository.provider.label}:${root.path}`;
+			let historyResult = this.storageService.get(historyKey, StorageScope.WORKSPACE);
+			if (historyResult) {
+				this.history = JSON.parse(historyResult);
+			}
+		}
+	}
+	private readonly _onDidClearHistory = new Emitter<void>();
+	readonly onDidClearHistory: Event<void> = this._onDidClearHistory.event;
```

Comment:
Why do we have this event at all? Seems like it's not connected to anything.

Reply:
Also, it's important to keep the code style: remember to place fields and properties together, before the constructor.